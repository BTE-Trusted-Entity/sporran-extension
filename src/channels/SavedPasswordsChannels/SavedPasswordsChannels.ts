import { BrowserChannel } from '../base/BrowserChannel/BrowserChannel';
import { channelsEnum } from '../base/channelsEnum';

interface SavedPassword {
  password: string;
  timestamp: number;
}

const savedPasswords: Record<string, SavedPassword> = {};

interface SavePasswordInput {
  password: string;
  address: string;
}

export const savePasswordChannel = new BrowserChannel<SavePasswordInput>(
  channelsEnum.savePassword,
);

export async function savePassword({
  address,
  password,
}: SavePasswordInput): Promise<void> {
  savedPasswords[address] = { password, timestamp: Date.now() };
}

export const getPasswordChannel = new BrowserChannel<
  string,
  string | undefined
>(channelsEnum.getPassword);

export async function getPassword(
  address: string,
): Promise<string | undefined> {
  return savedPasswords[address]?.password;
}

export const forgetPasswordChannel = new BrowserChannel<string>(
  channelsEnum.forgetPassword,
);

export async function forgetPassword(address: string): Promise<void> {
  delete savedPasswords[address];
}

export const hasSavedPasswordsChannel = new BrowserChannel<void, boolean>(
  channelsEnum.hasSavedPasswords,
);

export async function hasSavedPasswords(): Promise<boolean> {
  return Object.values(savedPasswords).length > 0;
}

export const forgetAllPasswordsChannel = new BrowserChannel(
  channelsEnum.forgetAllPasswords,
);

export async function forgetAllPasswords(): Promise<void> {
  for (const password in savedPasswords) {
    delete savedPasswords[password];
  }
}

const saveDuration = 15 * 60 * 1000;
const intervalDuration = 1 * 60 * 1000;

function checkExpiredPasswords(): void {
  const oldestPossible = Date.now() - saveDuration;
  for (const password in savedPasswords) {
    const { timestamp } = savedPasswords[password];
    if (timestamp < oldestPossible) {
      delete savedPasswords[password];
    }
  }
}

export function schedulePasswordsCheck(): void {
  setInterval(checkExpiredPasswords, intervalDuration);
}
