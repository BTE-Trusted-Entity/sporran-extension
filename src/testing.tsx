export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
