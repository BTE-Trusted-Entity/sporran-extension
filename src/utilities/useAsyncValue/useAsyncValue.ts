import { useEffect, useState } from 'react';

import { exceptionToError } from '../exceptionToError/exceptionToError';

export function useAsyncValue<Input extends Array<unknown>, Output>(
  fetcher: (...args: Input) => Promise<Output>,
  args: Input,
): Output | undefined {
  const [data, setData] = useState<Output>();
  const [error, setError] = useState<Error>();

  useEffect(() => {
    (async () => {
      try {
        setData(await fetcher(...args));
      } catch (exception) {
        console.error(exception);
        setError(exceptionToError(exception));
      }
    })();
  }, args); // eslint-disable-line react-hooks/exhaustive-deps

  if (error) {
    throw error;
  }

  return data;
}
