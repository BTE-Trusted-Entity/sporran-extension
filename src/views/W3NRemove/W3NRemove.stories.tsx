import { Meta } from '@storybook/react';

import { moreIdentitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';

import { W3NRemove } from './W3NRemove';

export default {
  title: 'Views/W3NRemove',
  component: W3NRemove,
} as Meta;

export function NoRefund(): JSX.Element {
  return (
    <W3NRemove
      identity={identities['4tayr7qa5BoqQjbpDdVSkJNHYBCx9BZ2baf5fkJjFXehZKKe']}
    />
  );
}

export function Refund(): JSX.Element {
  return (
    <W3NRemove
      identity={identities['4q11Jce9wqM4A9GPB2z8n4K8LF9w2sQgZKFddhuKXwQ2Qo4q']}
    />
  );
}
