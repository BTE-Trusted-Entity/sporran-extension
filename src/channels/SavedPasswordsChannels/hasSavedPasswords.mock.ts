import { act } from '@testing-library/react';

import { hasSavedPasswordsChannel } from './SavedPasswordsChannels';

jest.mock('./SavedPasswordsChannels');
const hasSavedPasswordsPromise = Promise.resolve(false);
(hasSavedPasswordsChannel.get as jest.Mock).mockReturnValue(
  hasSavedPasswordsPromise,
);

export async function waitForHasSavedPasswords(): Promise<void> {
  await act(async () => {
    await hasSavedPasswordsPromise;
  });
}
