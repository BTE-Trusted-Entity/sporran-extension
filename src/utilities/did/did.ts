import {
  ConfigService,
  Did,
  DidDocument,
  DidEncryptionKey,
  DidUri,
  Utils,
} from '@kiltprotocol/sdk-js';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function isFullDid(did: DidUri | undefined): boolean {
  if (!did) {
    // could be a legacy identity without DID
    return false;
  }
  return Did.parse(did).type === 'full';
}

export async function getFullDidDocument(did: DidUri): Promise<DidDocument> {
  const api = ConfigService.get('api');
  return Did.linkedInfoFromChain(await api.call.did.query(Did.toChain(did)))
    .document;
}

export function useFullDidDocument(did: DidUri): DidDocument | undefined {
  return useAsyncValue(getFullDidDocument, [did]);
}

export async function getDidDocument(did: DidUri): Promise<DidDocument> {
  return isFullDid(did)
    ? await getFullDidDocument(did)
    : Did.parseDocumentFromLightDid(did);
}

export function getDidEncryptionKey({
  keyAgreement,
}: DidDocument): DidEncryptionKey {
  if (!keyAgreement || !keyAgreement[0]) {
    throw new Error('encryptionKey is not defined somehow');
  }
  return keyAgreement[0];
}

export function parseDidUri(did: DidUri): ReturnType<typeof Did.parse> & {
  fullDid: DidUri;
} {
  const parsed = Did.parse(did);
  const fullDid = parsed.type === 'full' ? did : Did.getFullDidUri(did);

  return {
    ...parsed,
    fullDid,
  };
}

export async function needLegacyDidCrypto(
  did: DidUri | undefined,
): Promise<boolean> {
  if (!did) {
    // DID was deactivated, no action needed.
    return false;
  }
  if (!isFullDid(did)) {
    return false;
  }

  try {
    const encryptionKey = getDidEncryptionKey(await getDidDocument(did));
    return (
      Utils.Crypto.u8aToHex(encryptionKey.publicKey) ===
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e'
    );
  } catch {
    // getDidDocument might throw if the DID is not on-chain anymore (removed, another endpoint),
    // no legacy crypto needed in that case
    return false;
  }
}
