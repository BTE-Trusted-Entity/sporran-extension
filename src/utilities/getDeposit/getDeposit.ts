import { ConfigService, Did, DidUri, KiltAddress } from '@kiltprotocol/sdk-js';

import BN from 'bn.js';

import { isFullDid } from '../did/did';
import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

interface DepositData {
  amount: BN;
  owner?: KiltAddress;
}

async function getDefaultDeposit() {
  const api = ConfigService.get('api');
  return { amount: api.consts.web3Names.deposit };
}

async function getDepositWeb3Name(
  did: DidUri | undefined,
): Promise<DepositData | undefined> {
  if (!did || !isFullDid(did)) {
    return getDefaultDeposit();
  }

  const api = ConfigService.get('api');

  const { web3Name } = Did.linkedInfoFromChain(
    await api.call.did.query(Did.toChain(did)),
  );

  if (!web3Name) {
    return getDefaultDeposit();
  }

  const data = await api.query.web3Names.owner(web3Name);

  if (data.isNone) {
    return;
  }
  return Did.depositFromChain(data.unwrap().deposit);
}

export function useDepositWeb3Name(
  did: DidUri | undefined,
): DepositData | undefined {
  return useAsyncValue(getDepositWeb3Name, [did]);
}

export async function getDepositDid(
  did: DidUri | undefined,
): Promise<DepositData | undefined> {
  const api = ConfigService.get('api');

  if (!did || Did.parse(did).type === 'light') {
    return { amount: api.consts.did.deposit };
  }

  return Did.depositFromChain(
    (await api.query.did.did(Did.toChain(did))).unwrap().deposit,
  );
}

export function useDepositDid(
  did: DidUri | undefined,
): DepositData | undefined {
  return useAsyncValue(getDepositDid, [did]);
}
