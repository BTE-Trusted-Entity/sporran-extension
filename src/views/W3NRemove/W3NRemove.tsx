import { FormEvent, Fragment, useCallback } from 'react';
import { generatePath, Link } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';

import { ConfigService, Did, DidDocument } from '@kiltprotocol/sdk-js';

import * as styles from './W3NRemove.module.css';

import { Identity } from '../../utilities/identities/types';

import {
  getIdentityCryptoFromSeed,
  getIdentityDid,
} from '../../utilities/identities/identities';
import { paths } from '../paths';

import { IdentitySlide } from '../../components/IdentitySlide/IdentitySlide';
import {
  PasswordField,
  usePasswordField,
} from '../../components/PasswordField/PasswordField';
import { KiltCurrency } from '../../components/KiltCurrency/KiltCurrency';
import {
  asKiltCoins,
  KiltAmount,
} from '../../components/KiltAmount/KiltAmount';
import { TxStatusModal } from '../../components/TxStatusModal/TxStatusModal';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { Stats } from '../../components/Stats/Stats';
import { useFullDidDocument } from '../../utilities/did/did';
import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';
import { useDepositWeb3Name } from '../../utilities/getDeposit/getDeposit';
import { useSubmitStates } from '../../utilities/useSubmitStates/useSubmitStates';
import { makeFakeIdentityCrypto } from '../../utilities/makeFakeIdentityCrypto/makeFakeIdentityCrypto';

async function getFee(fullDid?: DidDocument) {
  if (!fullDid) {
    return undefined;
  }

  const { address, keypair, sign } = makeFakeIdentityCrypto();

  const api = ConfigService.get('api');

  const authorized = await Did.authorizeTx(
    fullDid.uri,
    api.tx.web3Names.releaseByOwner(),
    sign,
    address,
  );

  return (await authorized.paymentInfo(keypair)).partialFee;
}

interface Props {
  identity: Identity;
}

export function W3NRemove({ identity }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const { address } = identity;
  const did = getIdentityDid(identity);

  const destination = generatePath(paths.identity.did.manage.start, {
    address,
  });

  const deposit = useDepositWeb3Name(did);

  const isDepositOwner = deposit?.owner === address;

  const fullDidDocument = useFullDidDocument(did);
  const fee = useAsyncValue(getFee, [fullDidDocument]);

  const { submit, modalProps, submitting, unpaidCosts } = useSubmitStates();

  const passwordField = usePasswordField();
  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();
      if (!fullDidDocument) {
        return;
      }

      const { keypair, seed } = await passwordField.get(event);
      const { sign } = await getIdentityCryptoFromSeed(seed);

      const api = ConfigService.get('api');

      const authorized = await Did.authorizeTx(
        fullDidDocument.uri,
        api.tx.web3Names.releaseByOwner(),
        sign,
        keypair.address,
      );
      await submit(keypair, authorized);
    },
    [passwordField, fullDidDocument, submit],
  );

  if (!fee || !fullDidDocument) {
    return null; // blockchain data pending
  }

  const total = deposit?.amount.sub(fee);

  return (
    <form
      onSubmit={handleSubmit}
      className={styles.container}
      autoComplete="off"
    >
      <h1 className={styles.heading}>{t('view_W3NRemove_heading')}</h1>
      <p className={styles.subline}>{t('view_W3NRemove_subline')}</p>

      <IdentitySlide identity={identity} options={false} />

      {!isDepositOwner && (
        <p className={styles.costs}>
          {t('view_W3NRemove_fee_as_total')}
          <KiltAmount amount={fee} type="costs" smallDecimals />
        </p>
      )}

      {isDepositOwner && total && (
        <Fragment>
          <p className={styles.costs}>
            {t('view_W3NRemove_total')}
            <KiltAmount amount={total} type="costs" smallDecimals />
          </p>

          <p className={styles.details}>
            {t('view_W3NRemove_deposit')}
            {asKiltCoins(deposit?.amount, 'funds')} <KiltCurrency />
            {t('view_W3NRemove_fee')}
            {asKiltCoins(fee, 'costs')} <KiltCurrency />
          </p>
        </Fragment>
      )}

      <p className={styles.explanation}>{t('view_W3NRemove_explanation')}</p>

      <PasswordField identity={identity} autoFocus password={passwordField} />

      <p className={styles.buttonsLine}>
        <Link to={paths.home} className={styles.cancel}>
          {t('common_action_cancel')}
        </Link>
        <button type="submit" className={styles.submit} disabled={submitting}>
          {t('common_action_sign')}
        </button>
        <output
          className={styles.errorTooltip}
          hidden={!unpaidCosts || Boolean(modalProps)}
        >
          {t('view_W3NRemove_insufficientFunds', [asKiltCoins(fee, 'costs')])}
        </output>
      </p>

      {modalProps && (
        <TxStatusModal
          {...modalProps}
          identity={identity}
          destination={destination}
          messages={{
            pending: t('view_W3NRemove_pending'),
            success: t('view_W3NRemove_success'),
            error: t('view_W3NRemove_error'),
          }}
        />
      )}

      <LinkBack />
      <Stats />
    </form>
  );
}
