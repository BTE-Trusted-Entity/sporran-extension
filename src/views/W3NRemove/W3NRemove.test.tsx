import { BalanceUtils, DidDocument } from '@kiltprotocol/sdk-js';

import { identitiesMock, render } from '../../testing/testing';
import { waitForGetPassword } from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels.mock';
import '../../components/useCopyButton/useCopyButton.mock';
import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';
import { useDepositWeb3Name } from '../../utilities/getDeposit/getDeposit';
import { useFullDidDocument } from '../../utilities/did/did';

import { W3NRemove } from './W3NRemove';

const identity =
  identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];

jest.mock('../../utilities/did/did');
jest.mocked(useFullDidDocument).mockReturnValue({} as DidDocument);

jest.mock('../../utilities/useAsyncValue/useAsyncValue');
jest.mocked(useAsyncValue).mockReturnValue(BalanceUtils.toFemtoKilt(0.01));

jest.mock('../../utilities/getDeposit/getDeposit');

describe('W3NRemove', () => {
  it('should show refund amount including deposit if promo was not used', async () => {
    jest.mocked(useDepositWeb3Name).mockReturnValue({
      owner: identity.address,
      amount: BalanceUtils.toFemtoKilt(2),
    });

    const { container } = render(<W3NRemove identity={identity} />);
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });

  it('should show only fee amount if promo was used', async () => {
    jest.mocked(useDepositWeb3Name).mockReturnValue({
      owner: '4some other deposit owner',
      amount: BalanceUtils.toFemtoKilt(2),
    });

    const { container } = render(<W3NRemove identity={identity} />);
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });
});
