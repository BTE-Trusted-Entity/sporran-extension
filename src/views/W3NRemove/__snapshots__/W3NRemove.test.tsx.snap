// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`W3NRemove should show only fee amount if promo was used 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Remove web3name
    </h1>
    <p
      class="subline"
    >
      Remove web3name from the KILT blockchain
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="costs"
    >
      Fee: 
      <span
        class="amount"
      >
        0
        <span
          class="fractional"
        >
          .0100
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <p
      class="explanation"
    >
      Please note: if you remove your web3name, it can be claimed by someone else.
    </p>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least 0.0100 KILTs to pay the fee
      </output>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;

exports[`W3NRemove should show refund amount including deposit if promo was not used 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Remove web3name
    </h1>
    <p
      class="subline"
    >
      Remove web3name from the KILT blockchain
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="costs"
    >
      Refund Amount: 
      <span
        class="amount"
      >
        1
        <span
          class="fractional"
        >
          .9900
        </span>
         
        <span
          aria-label="KILT Coin"
          class="coin"
        />
      </span>
    </p>
    <p
      class="details"
    >
       Deposit 
      2.0000
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
       - fee 
      0.0100
       
      <span
        aria-label="KILT Coin"
        class="coin"
      />
    </p>
    <p
      class="explanation"
    >
      Please note: if you remove your web3name, it can be claimed by someone else.
    </p>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
      <output
        class="errorTooltip"
        hidden=""
      >
        You need at least 0.0100 KILTs to pay the fee
      </output>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </form>
</div>
`;
